File: tools/retrieval_tool.py
import braintrust
from pinecone.grpc import PineconeGRPC as Pinecone
from pydantic import BaseModel
from openai import OpenAI
import voyageai
import os
from typing import List

project = braintrust.projects.create("Spotify")
 
 
class Args(BaseModel):
    query: str

class DocumentOutput(BaseModel):
    artist: str
    name: str
    text: str

class DocumentOutputs(BaseModel):
    documents: List[DocumentOutput]
 
def handler(query: str):
    
    #Setup Voyage and Pinecone clients
    vo = voyageai.Client(api_key=os.getenv('VOYAGE_API_KEY'))
    pc = Pinecone(api_key=os.getenv('PINECONE_API_KEY'))
    index = pc.Index("spotify-braintrust-lyrics")

    #Embed query and perform cosine similarity search to Pinecone, then create a dict of results
    xq = vo.embed(query, model="voyage-3", input_type='query').embeddings[0]
    res = index.query(xq, top_k=20, include_metadata=True)
    res = [{"artist": match['metadata']['artist'],"name": match['metadata']['name'], "text": match['metadata']['text'] } for match in res['matches']]
    
    #Grab the text from the results and rerank them using Voyage. Return the top 15 after reranking
    text = [match['text']for match in res]
    reranked_docs = vo.rerank(query, text, model='rerank-2', top_k=15)
    indices = [doc.index for doc in reranked_docs.results ]
    docs = [res[x] for x in indices]
    
    return docs
 


get_documents = project.tools.create(
    handler=handler,
    name="Get Documents",
    slug="get-documents_rerank",
    description="fetch documents from Pinecone based on a user query",
    parameters=Args,
    returns=DocumentOutputs,
    if_exists="replace"
)
 
project.prompts.create(
    model="gpt-4o",
    name="Get related documents from Pinecone",
    slug="get-documents-prompt",
    messages=[
        {
            "role": "system",
            "content": "You are a musical historian. Upon receiving a query, you will generate a list of documents that are related to the query.",
        },
        {
            "role": "user",
            "content": "{{{query}}}",
        },
    ],
    tools=[get_documents],
    if_exists="replace"
)

File: tools/retrieval_tool.ts
import { VoyageAIClient } from 'voyageai';
import { Pinecone, ScoredPineconeRecord } from '@pinecone-database/pinecone';
import * as braintrust from "braintrust";
import { PROJECT_NAME } from '@/lib/constants';
import { z } from "zod";

const project = braintrust.projects.create({ name: PROJECT_NAME });

interface DocumentOutput {
  artist: string;
  name: string;
  text: string;
}

async function handler(query: string): Promise<DocumentOutput[]> {
  // Setup Voyage and Pinecone clients
  const vo = new VoyageAIClient({apiKey: process.env.VOYAGE_API_KEY!});
  const pc = new Pinecone({apiKey: process.env.PINECONE_API_KEY!});
  const index = pc.Index('spotify-braintrust-lyrics');

  // Embed query and perform cosine similarity search to Pinecone
  const embedding = await vo.embed({
    model: 'voyage-3',
    input: [query],
    inputType: 'query'
  });  
  const xq = embedding.data?.[0]?.embedding as number[];
  
  const queryResponse = await index.query({
    vector: xq,
    topK: 20,
    includeMetadata: true
  });

  const matches = queryResponse.matches as ScoredPineconeRecord[];
  console.log(queryResponse.matches)
  
  const res = matches.map(match => ({
    artist: match.metadata?.artist as string,
    name: match.metadata?.name as string,
    text: match.metadata?.text as string
  } ));

  // Rerank using Voyage
  const texts = res.map(match => match.text) as string[];
  const reranked = await vo.rerank({
    query: query,
    documents: texts,
    model: 'rerank-2',
    topK: 15
  });

  const indices = reranked.data?.map(doc => doc.index) as number[];
  const docs = indices.map(i => res[i]);

  return docs;
}

project.tools.create({
  handler: handler,
  name: "Typescript Pinecone Search",
  slug: "get-pinecone-docs-ts",
  description:
    "a way to retrieve and rerank documents from pinecone.",
  parameters: z.string(),
  ifExists: "replace",
});


export { handler };